src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseExplicitTypeCodeFixProvider.cs

```diff
@@ -74,42 +80,68 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             if (declarationContext is VariableDeclarationSyntax varDecl)
             {
                 typeSyntax = varDecl.Type;
+                variableIdentifier = varDecl.Variables.First().Identifier;
             }
             else if (declarationContext is ForEachStatementSyntax forEach)
             {
                 typeSyntax = forEach.Type;
+                variableIdentifier = forEach.Identifier;
             }
             else if (declarationContext is DeclarationExpressionSyntax declarationExpression)
             {
                 typeSyntax = declarationExpression.Type;
-                if (declarationExpression.Designation.IsKind(SyntaxKind.ParenthesizedVariableDesignation, out ParenthesizedVariableDesignationSyntax variableDesignation))
+                if (declarationExpression.Designation.IsKind(SyntaxKind.ParenthesizedVariableDesignation, out ParenthesizedVariableDesignationSyntax? variableDesignation))
                 {
                     parensDesignation = variableDesignation;
                 }
             }
             else
             {
-                throw ExceptionUtilities.UnexpectedValue(declarationContext.Kind());
+                throw ExceptionUtilities.UnexpectedValue(declarationContext?.Kind());
             }
 
             if (parensDesignation is null)
             {
                 typeSyntax = typeSyntax.StripRefIfNeeded();
 
+                var newTypeSymbol = semanticModel.GetTypeInfo(typeSyntax, cancellationToken).ConvertedType;
+                RoslynDebug.AssertNotNull(newTypeSymbol);
+
+                if (newTypeSymbol.NullableAnnotation == NullableAnnotation.Annotated && variableIdentifier.HasValue && variableIdentifier.Value.Parent is not null)
+                {
+                    var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+                    // It's possible that the var shouldn't be annotated nullable, check assignments to the variable and 
+                    // determine if it needs to be null
+                    var encapsulatingNode = node.FirstAncestorOrSelf<SyntaxNode>(n => syntaxFacts.IsMethodBody(n) || n is CompilationUnitSyntax);
+                    RoslynDebug.AssertNotNull(encapsulatingNode);
+
+                    var operationScope = semanticModel.GetOperation(encapsulatingNode, cancellationToken);
+                    var declSymbol = semanticModel.GetDeclaredSymbol(variableIdentifier.Value.Parent, cancellationToken);
+
+                    if (NullableHelpers.IsSymbolAssignedPossiblyNullValue(semanticModel, operationScope, declSymbol) == false)
```

Comment:
IsSymbolAssignedPossiblyNullValue will never return null in this case, because the var always at least one assignment, namely the initializer? I was tripped up trying to think about that carefully but then maybe realized it didn't matter.

Reply:
if it can't ever return null, i would love if we could contract/assert that.

Reply:
Changed so that the usage here doesn't return null